enhancements/security/security-profiles-operator.md

```diff
@@ -0,0 +1,636 @@
+been developed. The intent is that selinuxd allows one to simply drop a file
+on a pre-determined directory to install a SELinux policy, and that policy will
+get picked up and installed by the daemon. Thus, the SPOD instance simply
+places the policy into a directory, the selinuxd instance picks it up
+and handles the installation and tracking of policies.
+
+A `SelinuxProfile` object looks as follows:
+
+```yaml
+---
+apiVersion: security-profiles-operator.x-k8s.io/v1alpha1
+kind: SelinuxProfile
+metadata:
+  name: errorlogger
+  namespace: my-namespace
+spec:
+  policy: |
+    (blockinherit container)
+    (allow process var_log_t ( dir ( open read getattr lock search ioctl add_name remove_name write ))) 
+    (allow process var_log_t ( file ( getattr read write append ioctl lock map open create  ))) 
+    (allow process var_log_t ( sock_file ( getattr read write append open  )))
+```
+
+It is mostly a wrapper around [SELinux's CIL language](https://github.com/SELinuxProject/cil)
+with the main block and naming handled by the operator. Naming of policies
+as implemented by the daemon is as follows `<object name>_<object namespace>`. Thus,
+we're able to distinguish between policies installed in different namespaces.
+
+To ease usability and allow for developers to easily bind profiles to workloads,
+an object called `ProfileBinding` was created. The intent is for the binding to
+declare what profile goes to a ceratain image. The binding is then enforced by a
+webhook which mutates the pod definition to add the appropriate profile
+to its security context.
+
+The `ProfileBinding` object looks as follows:
+
+```yaml
+---
+apiVersion: security-profiles-operator.x-k8s.io/v1alpha1
+kind: ProfileBinding
+metadata:
+  name: profile-binding
+spec:
+  profileRef:
+    kind: SeccompProfile
+    name: profile-allow-unsafe
+  image: nginx:1.19.1
+```
+
+This would help manage security profiles and match them to workloads
+accross the application's lifecycle. Where one could upgrade their
+workload, and create a binding that matches the new application's image
+tag.
+
+Finally, to aide developers in writing profiles, there is a `ProfileRecording`
+object. The object would look as follows:
+
+```yaml
+---
+apiVersion: security-profiles-operator.x-k8s.io/v1alpha1
+kind: ProfileRecording
+metadata:
+  # The name of the Recording is the same as the resulting `SeccompProfile` CRD
+  # after reconciliation.
+  name: test-recording
+spec:
+  kind: SeccompProfile
+  podSelector:
+    matchLabels:
+      app: alpine
+```
+
+The operator automates the recording of profiles, and, when the
+workload is done, it's able to output a ready-to-use profile
+with the privileges that the application needs.
+
+The `SeccompProfile` objects rely on the [OCI Seccomp BPF Hook](https://github.com/containers/oci-seccomp-bpf-hook)
+to do the actual recording. So we'd need to ensure this hook is available in RHCOS
```

Comment:
Having this be an extension seems OK, but on the other hand if it is basically a no-op unless the annotation is present, then perhaps it's simplest to ship this by default (e.g. as part of crio?).  I don't have a really strong opinon.

Reply:
Having it enabled by default would be quite nice for usability IMO.

Reply:
How big is this package? If it's just a couple of KB, making an extension for that would probably not be worth it.

Reply:
> Having this be an extension seems OK, but on the other hand if it is basically a no-op unless the annotation is present, then perhaps it's simplest to ship this by default (e.g. as part of crio?). I don't have a really strong opinon.

I really don't know enough about crio to decide if having `oci-seccomp-bpf-hook` part of crio or a crio dependency makes sense. 

Just a heads-up: I was testing `oci-seccomp-bpf-hook` on OCP earlier this week and at the moment, `bcc` (a dependency of  `oci-seccomp-bpf-hook`) requires bcc-tools for some reason. This dependency is being removed, but if you try installing the packages from 8.4, you'll get  bcc-tools which brings some python deps and kernel-devel. You'll also want the hook from 8.5 because there was a bogus crun dependency. With these two deps removed, we're down to:

- oci-seccomp-bpf-hook (1MB)
- bcc (676kB)
- clang-libs (23MB)
- libbpf (101kB)
- llvm-libs (21MB)

If RHCOS can ship these by default, even better, but an extension would be fine by me. As said, we need to have the dependencies trimmed first in Fedora and RHEL, though. All details are in https://issues.redhat.com/browse/CMP-927